### ISSUE TITLE
Support Spring Security @Secured annotation

### ISSUE THREAD

**Issue by cbornet** [LogNet/grpc-spring-boot-starter#61]
Annotating a server implementation method with `@Secured` annotation currently doesn't have any effect.
It would be nice if the `@Secured` annotation is supported and a `Status.PERMISSION_DENIED` error is sent if the `Authentication` doesn't contain the proper authority.

```java
@Component
public class AuthenticationInterceptor implements ServerInterceptor {
    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> serverCall, Metadata metadata, ServerCallHandler<ReqT, RespT> serverCallHandler) {
        Authentication authentication = new UsernamePasswordAuthenticationToken("user", "user",
            Collections.singletonList(new SimpleGrantedAuthority(AuthoritiesConstants.USER));
        SecurityContextHolder.getContext().setAuthentication(authentication);
        return serverCallHandler.startCall(serverCall, metadata);  
    }
}
```
```java
@GRpcService(interceptors = { LogInterceptor.class, AuthenticationInterceptor.class })
public class GreeterService extends GreeterGrpc.GreeterImplBase {
    @Override
    @Secured({ "ROLE_ADMIN" })
    public void sayHello(GreeterOuterClass.HelloRequest request, StreamObserver<GreeterOuterClass.HelloReply> responseObserver) {
        String message = "Hello " + request.getName();
        final GreeterOuterClass.HelloReply.Builder replyBuilder = GreeterOuterClass.HelloReply.newBuilder().setMessage(message);
        responseObserver.onNext(replyBuilder.build());
        responseObserver.onCompleted();
        log.info("Returning " +message);
    }
}
```
then `stub.sayHello` should fail with `Status.PERMISSION_DENIED`

**Comment by jorgheymans** [comment]
probably same as #41 ?

**Comment by cbornet** [comment]
Maybe related but there is also the need to translate the exception. An exception translator like Spring's `ControllerAdvice` would be awesome

**Comment by pagrus7** [comment]
Out of interest I've put together [an example](https://github.com/pagrus7/grpc-spring-boot-starter/commit/15ab3c0114f821bf850039fef479b67c2d88db86) of Secured annotation (which just works) and a basic exception-to-status translation interceptor.

I don't think that would be sufficient to cover a meaningful scenario though. Just curious, where the authentication will be coming from in the intended use case?

**Comment by alexleigh** [comment]
I've also put together [a standalone example](https://github.com/revinate/grpc-spring-security-demo/commit/8af30c9627507580fabf608ac94819442549f617) of method-based security with gRPC (using pre-post-annotations instead of secured-annotations) with an exception-to-status translator, using Authorization metadata for Basic Auth credentials and JWT tokens. It works pretty well for our production use-case, though the default ThreadLocal security context storage is less than ideal.

**Comment by cbornet** [comment]
I guess I should try again because `spring.aop.proxy-target-class=true` didn't work  for me when I tried it ...

**Comment by alexleigh** [comment]
I've also had weird issues with the method security AOP interceptors not applying to the gRPC implementation class at times. In fact, in my standalone example, if I remove spring-jdbc from runtime dependencies, the AOP stops applying.

**Comment by pagrus7** [comment]
@alexleigh I think you are hitting another flavour of the autoproxying issue. By default spring boot 1.x would use a a dynamic proxies, but with spring-jdbc and spring-tx pulled into the classpath (the autoconfiguration)[https://github.com/spring-projects/spring-boot/blob/dac7d9c17449b286d31b04dba8385ee484552b53/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/TransactionAutoConfiguration.java#L94] kicks in and configures cglib proxies instead.

I opened a (PR)[https://github.com/revinate/grpc-spring-security-demo/pull/1] which explicitly specifies `proxyTargetClass = true` on `@EnableGlobalMethodSecurity` annotation. Security then still works without spring-boot-starter-jdbc.

**Comment by alexleigh** [comment]
That's really good to know. Thanks for the help!

**Comment by ST-DDT** [comment]
See also https://github.com/grpc/grpc-java/issues/4970

**Comment by cbornet** [comment]
@ST-DDT I've seen your discussion on the grpc-java issue. Do you know how to proceed ?

**Comment by ST-DDT** [comment]
In order for SpringSecurity to work you need two things:
1. ServerInterceptors (See the [example](https://github.com/pagrus7/grpc-spring-boot-starter/commit/15ab3c0114f821bf850039fef479b67c2d88db86) linked above)
   * With Exception mapping
   * With SecurityContext builder/Authentication
2. An annotated GrpcService impl
   * Which has either an interface with ALL methods (created and maintained by yourself, or somehow generated at compile time)
   * Or you use `@EnableGlobalMethodSecurity(proxyTargetClass = true)` and be okay with one line of warning per service class during each startup.

If you get an `INTERNAL` status response, then you have an issue with the first one.
If you get an `UNAVAILABLE` status response, then you have an issue with the second one.

As an alternative you could avoid the security annotation and use another server interceptor to map the access checks yourself. Using something like similar to a `Map<MethodDescriptor, AccessDecider>`.

**Comment by jvmlet** [comment]
@cbornet , @ST-DDT , have you seen [this ](https://eng.revinate.com/2017/11/07/grpc-spring-security.html) ? Hi  is using this starter and implemented interceptors to integrate spring security together with authorization.
I've asked the author if he want to contribute to this repo... still no answer.

**Comment by ST-DDT** [comment]
That example is somewhat better than the above mentioned, as it used an AuthenticationManager. But it also has a few drawbacks:

* It hasn't been contributed to this project and is no standalone library (-> copy & paste)
* It uses a different license (-> so you have to actually add the license statements to those classes/wrapper library (not that big of a problem though))
* Analyses the cause stack which you might not want, especially if you have other ExceptionMappers
* Lacks in code documentation/javadocs

Don't get me wrong, it is good code, but puts some obstacle in your way to use and maintain it.

**Comment by ST-DDT** [comment]
**WARNING:** The [security demo by revinate](https://github.com/revinate/grpc-spring-security-demo) and the [security example by pagrus7](https://github.com/pagrus7/grpc-spring-boot-starter/commit/15ab3c0114f821bf850039fef479b67c2d88db86) are vulnerable to [concurrency issues](https://github.com/revinate/grpc-spring-security-demo/issues/2).

You might see the following issues:
* Unauthenticated users can execute calls
* Authenticated requests get rejected due to `UNAUTHENTICATED`
* Authenticated requests might show a different username

During my tests this issues only occurred in very few cases (aka hard to reproduce).
Its easier to reproduce the bug with many concurrent calls by different users (including unauthenticated ones) and maybe a busy CPU (to force the thread context switches).

You can avoid that issue by rewriting the authenticating interceptors to work similar to [`grpc-contexts`](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/Contexts.java#L44).
Or have a look at my PR (https://github.com/yidongnan/grpc-spring-boot-starter/pull/126) which tries to fix that issue for a different grpc-spring-boot library.

See also [this SO question/answer](https://stackoverflow.com/questions/34905499/grpc-java-pass-data-from-server-interceptor-to-rpc-service-call)

-----------------------------------

That issue is not related to this repository. It's only related to the mentioned security demo and all variants that work similar.

**Comment by alex-lzl** [comment]
> **WARNING:** The [security demo by revinate](https://github.com/revinate/grpc-spring-security-demo) and the [security example by pagrus7](https://github.com/pagrus7/grpc-spring-boot-starter/commit/15ab3c0114f821bf850039fef479b67c2d88db86) are vulnerable to [concurrency issues](https://github.com/revinate/grpc-spring-security-demo/issues/2).

THANK YOU SO MUCH!!! I just ran into the concurrency issue mentioned above and super lucky found your post. Also just in time for the release of net.devh:grpc-spring-boot-starter:2.2.0.RELEASE.

**Comment by majusko** [comment]
@cbornet  , @ST-DDT , @jvmlet 

Probably not interested anymore, but if so: 

I created a simple JWT Spring Boot Starter extending this library from LogNet: https://github.com/majusko/grpc-jwt-spring-boot-starter

Already using it in few production projects so it will be definitely supported. Also, feel free to contribute ;)  

Simple usage showcase here:
https://github.com/majusko/grpc-example/blob/master/src/main/kotlin/io/github/majusko/grpc/example/ServerExample.kt

**Comment by jvmlet** [comment]
Thanks for sharing, @majusko. I'll definitely have a look

**Comment by balchua** [comment]
Is there some plan on supporting spring security? Thanks.

**Comment by jvmlet** [comment]
It's better later than never :-), implemented in `4.0.0`
